File: packages/core_crypto/android/src/main/kotlin/com/ktwallet/core_crypto/WalletIdValidation.kt
package com.ktwallet.core_crypto

internal class InvalidWalletIdException : Exception("invalid wallet id")

private val walletIdPattern = Regex("^[A-Za-z0-9_-]{1,64}$")

/**
 * Native trust-boundary validation for identifiers used as Keystore aliases
 * and entropy-blob file names. Dart performs the same check, but native code
 * must remain safe when invoked directly through a MethodChannel or test host.
 */
internal fun requireValidWalletId(value: Any?): String {
    val walletId = value as? String ?: throw InvalidWalletIdException()
    if (!walletIdPattern.matches(walletId)) throw InvalidWalletIdException()
    return walletId
}

File: packages/core_crypto/ios/core_crypto/Sources/core_crypto/WalletIdValidation.swift
import Foundation

enum WalletIdValidationError: Error {
  case invalid
}

/// Native trust-boundary validation for Keychain account identifiers. Dart
/// performs the same check, but direct MethodChannel calls must fail closed.
func requireValidWalletId(_ value: Any?) throws -> String {
  guard let walletId = value as? String,
    (1...64).contains(walletId.utf8.count),
    walletId.utf8.allSatisfy({ byte in
      (byte >= 0x41 && byte <= 0x5A)
        || (byte >= 0x61 && byte <= 0x7A)
        || (byte >= 0x30 && byte <= 0x39)
        || byte == 0x5F
        || byte == 0x2D
    })
  else {
    throw WalletIdValidationError.invalid
  }
  return walletId
}
